rockcraft/oci.py

```diff
@@ -0,0 +1,205 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2021 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""OCI image manipulation helpers."""
+
+
+import logging
+import os
+import shutil
+import subprocess
+import tarfile
+from dataclasses import dataclass
+from pathlib import Path
+from typing import List
+
+from rockcraft import ui
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class Image:
+    """A local OCI image.
+
+    :param image_name: The name of this image in ``name:tag`` format.
+    :param path: The path to this image in the local filesystem.
+    """
+
+    image_name: str
+    path: Path
+
+    @classmethod
+    def from_docker_registry(cls, image_name: str, *, image_dir: Path) -> "Image":
+        """Obtain an image from a docker registry.
+
+        :param image_name: The image to retrieve, in ``name:tag`` format.
+        :param image_dir: The directory to store local OCI images.
+
+        :returns: The downloaded image.
+        """
+        ui.emit.progress(f"Retrieving base {image_name}")
+
+        image_dir.mkdir(parents=True, exist_ok=True)
+        image_target = image_dir / image_name
+        _copy_image(f"docker://{image_name}", f"oci:{image_target}")
+
+        ui.emit.message(f"Retrieved base {image_name}")
+
+        return cls(image_name=image_name, path=image_dir)
+
+    def copy_to(self, image_name: str, *, image_dir: Path) -> "Image":
```

Comment:
This gets rid of no-self-use
```suggestion
    @staticmethod
    def copy_to(self, image_name: str, *, image_dir: Path) -> "Image":
```